manticore/core/cpu/abstractcpu.py

```diff
@@ -706,47 +724,75 @@ def execute(self):
         if issymbolic(self.PC):
             raise ConcretizeRegister(self, 'PC', policy='ALL')
 
-        if not self.memory.access_ok(self.PC,'x'):
+        if not self.memory.access_ok(self.PC, 'x'):
             raise InvalidMemoryAccess(self.PC, 'x')
 
         self.publish('will_decode_instruction', self.PC)
 
-        instruction = self.decode_instruction(self.PC)
-        self._last_pc=self.PC
-
-        self.publish('will_execute_instruction', instruction)
-
-        if instruction.address != self.PC:
-            return 
-
-        name = self.canonicalize_instruction_name(instruction)
-
-        def fallback_to_emulate(*operands):
-            text_bytes = ' '.join('%02x'%x for x in instruction.bytes)
-            logger.info("Unimplemented instruction: 0x%016x:\t%s\t%s\t%s",
-                    instruction.address, text_bytes, instruction.mnemonic,
-                    instruction.op_str)
-
-            self.publish('will_emulate_instruction', instruction)
-
-            self.emulate(instruction)
+        insn = self.decode_instruction(self.PC)
+        self._last_pc = self.PC
 
-            self.publish('did_emulate_instruction', instruction)
+        self.publish('will_execute_instruction', insn)
 
-        implementation = getattr(self, name, fallback_to_emulate)
+        # FIXME (theo) why just return here?
+        if insn.address != self.PC:
+            return
 
         if logger.level == logging.DEBUG :
-            logger.debug(self.render_instruction())
+            logger.debug(self.render_instruction(insn))
             for l in self.render_registers():
                 register_logger.debug(l)
-        
-        implementation(*instruction.operands)
-        self._icount+=1
 
+        self._insn_implementation(insn)
+        self._icount += 1
         self.publish('did_execute_instruction', instruction)
 
+    def _insn_implementation(self, insn):
+        name = self.canonicalize_instruction_name(insn)
+
+        def fallback_to_emulate(*operands):
```

Comment:
Perhaps make `fallback_to_emulate` a method on `Cpu` and provide an alternate implementation in `BinjaCpu`? Ideally, we shouldn't have Cpu-specific implementation here